Sprint 3/step 1

heyxesha opened the pull request.

heyxesha committed:
Добавлены страницы логина и регистрации.

heyxesha committed:
Добавила страницы для сброса пароля.

heyxesha committed:
Реализовала сброс пароля.

heyxesha committed:
Добавила страницу профиля.

heyxesha committed:
Добавила состояния авторизации.

heyxesha committed:
Добавила редактирование информации о пользователе.

heyxesha committed:
Реализовала защищенные маршруты.

heyxesha committed:
Перевела на формы.

heyxesha committed:
Реализовала карточку ингредиента.

heyxesha committed:
Добавила 404, запретила заказ неавторизованным пользователям, поправила пропсы.

heyxesha committed:
Поправила бесячие кавычки от автодополнения.

heyxesha committed:
Поправила ошибку с propTypes.

gennady-bars commented on src/pages/login.jsx:
```diff
@@ -0,0 +1,108 @@
+const LoginPage = () => {
+    const dispatch = useDispatch();
+    const navigate = useNavigate();
+    const [state, setState] = useState({
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/login.jsx:
```diff
@@ -0,0 +1,108 @@
+
+    return (
+        <FormPageWrapper showLoadingIndicator={ isLoginLoading } onSubmit={ onSubmit }>
+            <>
```
`<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку

gennady-bars commented on src/components/form-page-wrapper/form-page-wrapper.jsx:
```diff
@@ -0,0 +1,23 @@
+const FormPageWrapper = ({ children, showLoadingIndicator, onSubmit }) => {
+    return (
+        <PageWrapper showLoadingIndicator={ showLoadingIndicator }>
+            <form className={ `${ styles.Content } pb-10`}  onSubmit={ onSubmit }>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -1,18 +1,38 @@
+                const viewedIngredient = ingredients.find(item => item._id === params.id);
+                dispatch(setViewedIngredient(viewedIngredient));
+            } else {
+                dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/main/main.jsx:
```diff
@@ -0,0 +1,54 @@
+
+    const dispatch = useDispatch();
+    useEffect(() => {
+        dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/protected-route-element/protected-route-element.jsx:
```diff
@@ -0,0 +1,36 @@
+import { Navigate, useLocation } from 'react-router-dom';
+import { useSelector } from 'react-redux';
+
+const ProtectedRouteElement = ({ element, needAuthorization }) => {
```
**Можно лучше**

Если интересно, то можно сделать вот так универсальный компонент `ProtectedRoute` с переадресацией

```jsx


export default function ProtectedRoute({ children, anonymous = false }) {
  const isLoggedIn = useSelector((store) => store.user.isLoggedIn);

  const location = useLocation();
  const from = location.state?.from || '/';
  // Если разрешен неавторизованный доступ, а пользователь авторизован...
  if (anonymous && isLoggedIn) {
    // ...то отправляем его на предыдущую страницу
    return <Navigate to={ from } />;
  }

  // Если требуется авторизация, а пользователь не авторизован...
  if (!anonymous && !isLoggedIn) {
    // ...то отправляем его на страницу логин
    return <Navigate to="/login" state={{ from: location}}/>;
  }

  // Если все ок, то рендерим внутреннее содержимое
  return children;
}

```

gennady-bars commented on src/components/protected-route-element/protected-route-element.jsx:
```diff
@@ -0,0 +1,36 @@
+                state={
+                    {
+                        ...(location || {}),
+                        lastSecuredPage: window.location.pathname
```
`window.location`

**Можно лучше**


лучше использовать `location`, который с 8й строчки

Так будет надежнее и по-реактовски

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,59 +1,158 @@
+    const dispatch = useDispatch();
+    const navigate = useNavigate();
+    const location = useLocation();
+    const background = location.state?.background;
```
Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку
* Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно


но есть некоторые недочеты: 

* `<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку
* `dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

heyxesha committed:
Избавилась от <></>

heyxesha committed:
Перенесла запрос getIngredients() а уровень App, хоть и не считаю такой подход эффективным.
Опыт с ошибками оптимизациям говорит, что не нужно тянуть на страницу то, что она не использует.

heyxesha committed:
Немного переписала ProtectedRouteElement

heyxesha committed:
Перешла на useForm().

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.


* иконку нужно сделать больше в попапе и странице ингредиента https://disk.yandex.ru/i/xLYMUIqEg62cgw . В макете большая она

heyxesha committed:
Поправила размер ингредиента в попапе.
